### PR TITLE
Changes to use scikit-learn 0.24.0

### PR DESCRIPTION
**Pull request by R-Walmsley**
I think these is the minimal set of changes needed to address Issue #101. 

- imports _forest as private function 
- gives _parallel_build_trees_ an (unused) n_bootstrap_samples argument to match what it is passed from other functions 
- _parallel_build_trees_ does not change at all internally from previous versions of biobankAccelerometerAnalysis. As far as I can tell, this seems compatible with functions in 0.24.0 which call it - but may be risky when working with, and redefining, private functions. 

Feedback very welcome!

### CODE REVIEW

**Comment by angerhang** [comment]
The code looks good but has the model being retrained and uploaded? If we merge this PR beforehand, the model prediction will break as this happens to me when I tried to fix it :D

**Comment by angerhang** [comment]
Ideally, we should wait until the new model is uploaded and then merge this one.

**Comment by R-Walmsley** [comment]
Good point @angerhang...

Currently running the full training for the LOO analysis (I did a smaller set to test) and the final model, so will hopefully have an updated version of one model by tomorrow. The older models should follow shortly after.

**Comment by chanshing** [comment]
Thanks Rosemary, and agreed with Hang. Can we rebase it into one commit?

**Comment by R-Walmsley** [comment]
I think this is now ready to go? Have just updated the README and defaults with the new model names. Is there anywhere I've missed a default?  

@chanshing, if you don't mind, as it's not many commits (only 2 not made directly on GitHub), I'd rather not rebase this. It's had quite a messy history in which it's already been cherry-picked, and other things are written on top. So I'm worried about accidents if I force-push a rebased version. This is a good reminder to try  to push rebased commits in the future, though.

**Comment by chanshing** [comment]
> I think this is now ready to go? Have just updated the README and defaults with the new model names. Is there anywhere I've missed a default?
> 
> @chanshing, if you don't mind, as it's not many commits (only 2 not made directly on GitHub), I'd rather not rebase this. It's had quite a messy history in which it's already been cherry-picked, and other things are written on top. So I'm worried about accidents if I force-push a rebased version. But happy to discuss this! This is also a good reminder to try to push rebased commits in the future...

@R-Walmsley There are 2 pushes but 5 commits. It looks like the commits could be nicely squashed into one (in fact there are two commits with the same name). On my end I was able to squash them with the following steps:

### First look at `git log`

git_log_last5

### Rebase last 5 commits: `git rebase -i HEAD~5`
It will open an editor (probably vim if you haven't configured otherwise) with the following (sorry for the Spanish!):
rebasing
The first 5 lines show, from oldest to newest, the 5 commits at issue. We will squash the last 4 commits into the first (oldest).

### Squash last 4 commits into the first
Just change `pick` to `squash` or `s` for the last 4 newest commits. Recall that the list goes from oldest to newest.
squash_last4
Save and exit (`:wq` in vim). 

### Edit your new commit message 
After saving and exiting, the editor should reopen once again, this time to let you edit the new commit message: 
rebase_commit_message
As shown, there will be some default commit message already (again, sorry for the Spanish!), but you can delete this and write your own message:
rebase_custom_commit_message
Save and exit again (`:wq` in vim). You will return to the command line, hopefully with a success message:
rebase_success

### Recheck `git log`
Check again `git log` just to be sure. You should see the last 5 commits replaced by your newly created commit. 
It now looks like you have done everything in one go. You rewrote history to make you look like a badass :sunglasses: 
gitlog_after

**Comment by chanshing** [comment]
> and other things are written on top.

@R-Walmsley If I understand correctly, you currently have files with changes that are not yet committed. In that case, one solution is to run `git stash`, which will undo all the changes since HEAD temporarily but also stash them. Then you can do the rebasing as above. Once done, you can recover the changes by running `git stash apply`.

**Comment by R-Walmsley** [comment]
TL;DR: Using --f scares me.  

@chanshing, my worry is a bit more subtle I think. 

I'm worried because: 
- This would change the commit hashes. These commits have already had two hashes, as they were cherry-picked. 
- In the time since writing these, I've committed local changes on top of them, which might one day be merged in (e.g. one to do with the features used, which I needed to use along with the new version of this). I'm not sure how changing the hash will complicate any eventual merge of those, and I don't like doing things I don't understand with git. 

It's probably fine and nothing will break. But given the consequences of not doing it is only more commits than needed, that seemed to me the safe option.

I'm not arguing with the general principle of rebasing-before-merging, I just want to be confident we're doing it safely.

**Comment by chanshing** [comment]
> TL;DR: Using --f scares me.
> 
> @chanshing, my worry is a bit more subtle I think.
> 
> I'm worried because:
> 
> * This would change the commit hashes. These commits have already had two hashes, as they were cherry-picked.
> * In the time since writing these, I've committed local changes on top of them, which might one day be merged in (e.g. one to do with the features used, which I needed to use along with the new version of this). I'm not sure how changing the hash will complicate any eventual merge of those, and I don't like doing things I don't understand with git.
> 
> It's probably fine and nothing will break. But given the consequences of not doing it is only more commits than needed, that seemed to me the safe option.
> 
> I'm not arguing with the general principle of rebasing-before-merging, I just want to be confident we're doing it safely.

Oh I see. For future reference, you can create a new branch, branching out from your `branch-under-review` to keep working and committing stuff without stepping over the `branch-under-review`. Otherwise, if there were an actual revision it would mean you'd have to go back several commits to fix it, and even then I don't know how I'd push that branch without the newer unrelated commits :thinking: 

If it makes things easier, would you like me to try pushing the squashed history to `R-Walmsley:add-sklearn-changes-only`? It's something I am keen to try. Note that your local laptop/computer repo will remain unchanged.

**Comment by R-Walmsley** [comment]
> Oh I see. For future reference, you can create a new branch, branching out from your `branch-under-review` to keep working and committing stuff without stepping over the `branch-under-review`. Otherwise, if there were an actual revision it would mean you'd have to go back several commits to fix it, and even then I don't know how I'd push that branch without the newer unrelated commits 🤔

^This is definitely the right workflow in general. Let's write up a brief summary of this workflow so we don't have to use our brains in the future?

You're welcome to try pushing to it. I think that's still vulnerable to the worry I had about what would happen when merging anything built on top of those commits, because it will still change the hashes associated with these changes. But if I squash commit any future merge , I think git is clever enough to cope...

**Comment by chanshing** [comment]
> ^This is definitely the right workflow in general. Let's write up a brief summary of this workflow so we don't have to use our brains in the future?

Yep! Let me search where I learned about this workflow and share it in Teams

> You're welcome to try pushing to it. I think that's still vulnerable to the worry I had about what would happen when merging anything built on top of those commits, because it will still change the hashes associated with these changes. But if I squash commit any future merge , I think git is clever enough to cope...

Fascinating, I was able to push to your repo :dancing_men: Loving Github even more.
I think you might have to rebase/merge your branch to master once this PR is incorporated, the same way as with any new changes to master. In the luckiest case, it'll be a fast-forward merge. In other cases, someone worked on the same file as you did and there will be a merge conflict to be resolved. In any case, let me know of any future issues and I'll try to help :cat:

**Comment by R-Walmsley** [comment]
I like this summary because it has an easy checklist: https://medium.com/singlestone/a-git-workflow-using-rebase-1b1210de83e5

**Comment by chanshing** [comment]
@R-Walmsley the checks are still failing. I think the final missing piece is to update downloadDataModels.sh

CI checks
Captura de pantalla de 2021-01-27 08-06-42

downloadDataModels.sh
Captura de pantalla de 2021-01-27 08-08-50

I will go ahead and fix this as part of the commit. CI saving the day :cat:

**Comment by R-Walmsley** [comment]
@chanshing - we need to download both the old models and the new ones, for back-compatibility. (Think this was agreed by @angerhang and @aidendoherty last week)

**Comment by R-Walmsley** [comment]
@chanshing - we also need the new versions of the other 2 models (i.e. `doherty-jan21` and `willetts-jan21`)

**Comment by chanshing** [comment]
> @chanshing - we also need the new versions of the other 2 models (i.e. `doherty-jan21` and `willetts-jan21`)

Oh ok, so should it be:

```bash
rm activityModels/doherty-may20.tar
rm activityModels/willetts-may20.tar
rm activityModels/walmsley-nov20.tar
rm activityModels/doherty-jan21.tar
rm activityModels/willetts-jan21.tar
rm activityModels/walmsley-jan21.tar
wget ${downloadDir}doherty-may20.tar -P activityModels/
wget ${downloadDir}willetts-may20.tar -P activityModels/
wget ${downloadDir}walmsley-nov20.tar -P activityModels/
wget ${downloadDir}doherty-jan21.tar -P activityModels/
wget ${downloadDir}willetts-jan21.tar -P activityModels/
wget ${downloadDir}walmsley-jan21.tar -P activityModels/
```
?

**Comment by chanshing** [comment]
accProcess.py run successfully now, but it seems there are differences in the epoch files

Captura de pantalla de 2021-01-27 08-22-39

Maybe some random seed difference in the new models?

**Comment by chanshing** [comment]
Ok so I run a diff to see what's the difference between the new epoch file and the old one. The good news is that the differences were not due to the new model. The bad news is that the differences were due to the recent merge 2fabe36a681fda4349b8eb4e213a9e8ac64affac that handled the NaN correlations, but we (I) forgot to update the epoch file to reflect that. In the current referenceFiles/epoch.csv there are two rows (4902 & 13670) where some of `corrxy` and `corryz` are NaN. After 2fabe36a681fda4349b8eb4e213a9e8ac64affac they are now 0.0.

So I believe this PR should be ready to go. @R-Walmsley do you want to push the button?
I will separately submit a PR to update the reference epoch file.

**Comment by R-Walmsley** [comment]
Thank you for your detective work @chanshing !
